### PR TITLE
Adding analysis of main ping submission and recording delays

### PR DESCRIPTION
**Pull request by chutten**
I measured and plotted the recording and submission delays of "main" pings across the major channels.

### CODE REVIEW

**Comment by harterrt** [comment]
Thanks, @chutten. I'll give this a review in the morning.

File: projects/ping_delays.kp/knowledge.md [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    {'application/channel': u'release',
+     'creationDate': u'2017-01-10T02:01:31.551Z',
+     'meta/Date': u'Tue, 10 Jan 2017 02:01:31 GMT',
+     'meta/Timestamp': 1484013691737682688L,
+     'meta/creationTimestamp': 1.484013691551e+18,
+     'payload/info/subsessionLength': 67}
+
+
+
+Quick normalization: ditch any ping that doesn't have a subsessionLength, creationTimestamp, or Timestamp:
```
Comment (harterrt) [inline projects/ping_delays.kp/knowledge.md]: What proportion of pings are we dropping here?

File: projects/ping_delays.kp/knowledge.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    plt.plot(sortd, ys)
+```
+
+```python
+def calculate_delays(p):
+    reporting_delay = p["payload/info/subsessionLength"]
+    
+    created = datetime.fromtimestamp(p["meta/creationTimestamp"] / 1000.0 / 1000.0 / 1000.0)
+    received = datetime.fromtimestamp(p["meta/Timestamp"] / 1000.0 / 1000.0 / 1000.0)
+    sent = datetime.fromtimestamp(mktime_tz(parsedate_tz(p["meta/Date"]))) if p["meta/Date"] is not None else received
```
Comment (harterrt) [inline projects/ping_delays.kp/knowledge.md]: How frequently is "meta/Date" missing? Maybe we should exclude them if it's a noticeable proportion of observations.

File: projects/ping_delays.kp/knowledge.md (hunk 1 above)
Comment (chutten) [inline projects/ping_delays.kp/knowledge.md]: Very few. The last time I did this, across some 3-4million pings I was dropping a couple dozen. So... less than 0.01%?

File: projects/ping_delays.kp/knowledge.md (hunk 2 above)
Comment (chutten) [inline projects/ping_delays.kp/knowledge.md]: Any time we get a ping from a build that doesn't have bug 1144778 (landed in 48). Clock skew is generally pretty small (under 1s most of the time. I would point to a reference but the last time I analyzed Date vs Timestamp I sadly didn't save the notebook), so I was okay with handling pre-48 numbers without adjusting for skew.

**Comment by harterrt** [comment]
:thumbsup:

**Pull request closed without merging**